### PR TITLE
Script for run saps locally

### PR DESCRIPTION
**Pull request by thiagoyeds**

### CODE REVIEW

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+
+echo "Checking args"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: remove print in line 3. Whenever someone needs such fine-grained debug, `bash -x` applies

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+if [ $# -ne 5 ];
+then
+	echo "Execution using 5 (five) arguments:"
+	echo "arg1 - usgs username."
+	echo "arg2 - usgs password."
+	echo "arg3 - image dataset in format landsat_X, where X (landsat number)."
+	echo "arg4 - image region in format PPPRRR, where P (path) and R (row)."
+	echo "arg5 - image date in format YYYY-MM-DD, where Y (year), M (month) and D (day)."
+	exit 1
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: make it more standard. e.g

echo "Usage: $0 usgs_username usgs_password [landsat_5 | landsat_7 | landsat_8] region_pathrow image_date"

Comment (thiagomanel) [reply src/test/resources/local-pipeline/run_local_pipeline]: add an exempla, omitting user and password in the script comments

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+fi
+USGS_USERNAME=$1
+USGS_PASSWORD=$2
+IMAGE_DATASET=$3
+IMAGE_REGION=$4
+IMAGE_DATE=$5
+
+echo "Checking docker"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: remove print in line 20

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+which docker &> log
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: I think it is nice to check docker install. kudos!

However, I think it is better to NOT help the user installing it. Just `exit 1`

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+if [ $? -ne 0 ];
+then
+	apt-get update &>> log
+	apt-get install docker-ce docker-ce-cli containerd.io &>> log
+fi
+
+echo "Declare name images"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: remove print line 28

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+IMAGE_NAME_INPUTDOWNLOADER="fogbow/inputdownloader:latest"
+IMAGE_NAME_WORKER="fogbow/worker:latest"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: image URLs should be args, ENV vars or read from a conf file.

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+echo "Pull docker containers"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: remove print line 32

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+docker pull $IMAGE_NAME_INPUTDOWNLOADER &>> log
+docker pull $IMAGE_NAME_WORKER &>> log
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: I did not like this redirection to this log file. maybe, the user of local-pipeline should take care of this

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+echo "Declare name containers"
+CONTAINER_NAME_INPUTDOWNLOADER="inputdownloader_container"
+CONTAINER_NAME_WORKER="worker_container"
+
+echo "Create share folder between containers and host"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: where is the shareD folder was created?

**Review by thiagomanel — CHANGES_REQUESTED** [review]
hi @ThiagoWhispher good job. the code is small, which is good.

I would suggest some modifications before accepts this PR (in addition to the comments I have already made):
- remove all the calls to `echo`
- try to use functions as you do in better languages. It was expecting to see a download() preprocess() and worker() functions, at least
- when those functions fail, abort the execution of the script
- to do redirect the out stream to a log file

File: src/test/resources/local-pipeline/run_local_pipeline (hunk 9 above)
Comment (thiagoyeds) [reply src/test/resources/local-pipeline/run_local_pipeline]: The shared folder was created in /tmp in both containers.

**Comment by thiagoyeds** [comment]
Hi @thiagomanel, thanks for the suggestions.
I made the modifications listed, could you check them?

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 10]
```diff
@@ -0,0 +1,92 @@
+#!/bin/bash
+
+if [ $# -ne 5 ];
+then
+  echo "Usage: usgs_username usgs_password [landsat_5 | landsat_7 | landsat_8] region_pathrow image_date"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: you missed `$0` before de first parameter

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+  echo "For example: ./run_local_pipeline <usgs_username> <usgs_password> landsat_5 215065 2011-10-18"
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: I think it is ok to have this as a comment using `#` instead of a print

File: src/test/resources/local-pipeline/run_local_pipeline (hunk 6 above)
Comment (thiagomanel) [reply src/test/resources/local-pipeline/run_local_pipeline]: you missed addressing above comment

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+function exit_handler {
+  if [[ $? -ne 0 ]];
+  then
+    exit 1
+  fi
+}
+
+function check_docker {
+  which docker
+  exit_handler
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline:41]: why this `exit_handler` is within this function and in other places, after the function?

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+function build_container_inputdownloader {
+  docker ps -a | grep "$CONTAINER_NAME_INPUTDOWNLOADER"
+  if [ $? -eq 0 ];
+  then
+    docker stop $CONTAINER_NAME_INPUTDOWNLOADER
+    docker rm $CONTAINER_NAME_INPUTDOWNLOADER
+  fi
+
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline:84]: add a comment explaining this. it is not crystal clear, so add a comment (remember your own struggle to understand some obscure SAPS/SEBAL code some days ago due to the missing comments/doc)

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+function build_container_worker {
+  docker ps -a | grep "$CONTAINER_NAME_WORKER"
+  if [ $? -eq 0 ];
+  then
+    docker stop $CONTAINER_NAME_WORKER
+    docker rm $CONTAINER_NAME_WORKER
+  fi
+  docker run -it -d --name $CONTAINER_NAME_WORKER -v saps-vol:/tmp $IMAGE_NAME_WORKER
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: add comment

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+function run_download {
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: run_input_download

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+  docker exec $CONTAINER_NAME_INPUTDOWNLOADER mkdir -p /tmp/input /home/ubuntu/results /home/ubuntu/metadata
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: we need comments explaining this.

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+  docker exec -e USGS_USERNAME=$USGS_USERNAME -e USGS_PASSWORD=$USGS_PASSWORD $CONTAINER_NAME_INPUTDOWNLOADER sh /home/ubuntu/run.sh $IMAGE_DATASET $IMAGE_REGION $IMAGE_DATE /tmp/input /tmp/input
+}
+
+function run_worker {
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline:128]: in most of the cases, it is not good to have this hardwired strings (I mean, the paths). the typical question is: what is that? replace them by proper variable (local or global)

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+  docker exec $CONTAINER_NAME_INPUTDOWNLAODER mv /home/ubuntu/results /home/ubuntu/metadata /tmp
+  docker exec $CONTAINER_NAME_WORKER mkdir -p /tmp/preprocess
+  docker exec $CONTAINER_NAME_WORKER bash /home/ubuntu/bin/run.sh /tmp/input /tmp/results /tmp/preprocess /tmp/metadata
+}
+
+check_docker
+
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline:167]: it is better to build each image just after pulling it (remember that, in the future, we might want to run the N first steps instead of all the pipeline)

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+pull_image_inputdownloader
+exit_handler
+
+pull_image_worker
+exit_handler
+
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline:176]: where is the pre-processing function?

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+build_container_worker
+exit_handler
+
+run_download
+exit_handler
+
+run_worker
+exit_handler
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: the SAPS pipeline has an assumption that each step of the pipeline returns a number to indicate error/success. I would be great that the local pipeline reports them is some way. I suggest you return them in a single line, e.g: -1 2 31, in the end of the script execution

**Review by thiagomanel — CHANGES_REQUESTED** [review]
Almost done.

File: src/test/resources/local-pipeline/run_local_pipeline (hunk 19 above)
Comment (thiagoyeds) [reply src/test/resources/local-pipeline/run_local_pipeline:176]: Pre-processing function not exists because running in inputdownloader container already make download and preprocess in script run.

Comment (thiagomanel) [reply src/test/resources/local-pipeline/run_local_pipeline:176]: yep. however, note that it is just the current pipeline. In the future, any user might add a new pre-processing phase and we want to avoid changing your code to reflect that change. So, add a hook to the pre-processing

**Comment by thiagoyeds** [comment]
Done @thiagomanel

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 21]
```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+
+#Usage: $0 usgs_username usgs_password [landsat_5 | landsat_7 | landsat_8] region_pathrow image_date
+#For example: ./run_local_pipeline <usgs_username> <usgs_password> landsat_5 215065 2011-10-18
+
+#Checking args
+if [ $# -ne 5 ]
+then
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline:7]: @ThiagoWhispher the idea is to:
- keep the "Usage: .." in the if case
- remove the "for example ..."from the if case (keep it as comment)

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+#Global variables
+#Exit code variable
+EXIT_CODE_PROCESS=
+
+#Declare name images
+IMAGE_NAME_INPUTDOWNLOADER="fogbow/inputdownloader:latest"
+IMAGE_NAME_WORKER="fogbow/worker:latest"
+
+#Declare name containers
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: explain here why do we need it

Comment (thiagomanel) [reply src/test/resources/local-pipeline/run_local_pipeline]: we use container name to manage container lifecycle

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+function check_docker {
+  which docker
+}
+
+function pull_image_inputdownloader {
+  docker pull $IMAGE_NAME_INPUTDOWNLOADER
+}
+
+function pull_image_preprocessor {
+  :
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: instead of this, refactor this function and the other pull functions to do nothing when the string is empty

Comment (thiagomanel) [reply src/test/resources/local-pipeline/run_local_pipeline]: the run function also should test if the name is empty

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+    #Check if inputdownloader container already exists
+    docker ps -a | grep "$CONTAINER_NAME_INPUTDOWNLOADER"
+    if [ $? -eq 0 ]
+    then
+      #Stop and remove existing inputdownloader container
+      docker stop $CONTAINER_NAME_INPUTDOWNLOADER
+      docker rm $CONTAINER_NAME_INPUTDOWNLOADER
+    fi
+    #Build a inputdownloader container from your image
+    docker run -it -d --name $CONTAINER_NAME_INPUTDOWNLOADER -v saps-vol:/tmp $IMAGE_NAME_INPUTDOWNLOADER
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: replace /tmp by a variable, such as INPUT_DOWNLOADER_SHARED_FOLDER

**Review by thiagomanel — CHANGES_REQUESTED** [review]
new batch

**Comment by thiagoyeds** [comment]
Done @thiagomanel , good ideas to make code easier to understand.

File: src/test/resources/local-pipeline/run_local_pipeline [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+build_container_preprocessor
+
+pull_image_worker
+build_container_worker
+
+run_input_download
+run_preprocessor
+run_worker
+
+echo $EXIT_CODE_INPUTDOWNLOADER $EXIT_CODE_PREPROCESSOR $EXIT_CODE_WORKER
```
Comment (thiagomanel) [inline src/test/resources/local-pipeline/run_local_pipeline]: almost done, @ThiagoWhispher 
i would favor returning the code in the run_* methods; so, instead of using a global $EXIT_CODE_* use a local variable in this way: `exit_code_inputdownloader=$(run_input_download)

Also, now, it seems better to me to have three code blocks: pull, build and run (inputdownload), then pre-proc, then worker ...

**Review by thiagomanel — CHANGES_REQUESTED** [review]
almost done

File: src/test/resources/local-pipeline/run_local_pipeline (hunk 25 above)
Comment (thiagoyeds) [reply src/test/resources/local-pipeline/run_local_pipeline]: @thiagomanel , then:
- pull_image_x, build_container_x and run_x, would be moved to a function called x (e.g. inputdownloader) and the same thing to others?
- the other modification would be something like: `local exit_code_inputdownloader=$ (inputdownloader), where inputdownloader is the new function mentioned above?

Comment (thiagomanel) [reply src/test/resources/local-pipeline/run_local_pipeline]: I think your approach is even better than mine, so do as you suggested

**Review by thiagomanel — CHANGES_REQUESTED** [review]
almost almost

**Comment by thiagoyeds** [comment]
Done @thiagomanel

**Review by thiagomanel — APPROVED** [review]
good job, @ThiagoWhispher
